### PR TITLE
HPCC-21321 Lsblk pipe cmd log stderr only on error

### PR DESCRIPTION
**Pull request by mckellyln**
Prevent stderr output from lsblk pipe->run cmd when cmd is successful.

Signed-off-by: Mark Kelly <mark.kelly@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:

Test suite and sample cmd that echoes to stdout and stderr and returns 0 and 1

### CODE REVIEW

**Comment by mckellyln** [comment]
@jakesmith please review.

File: system/jlib/jdebug.cpp [hunk 1]
```diff
@@ -1995,11 +1995,14 @@ class CExtendedStats  // Disk network and cpu stats
...
         Owned<IPipeProcess> pipe = createPipeProcess();
-        if (pipe->run("list disks", cmd, nullptr, false, true))
+        if (pipe->run("list disks", cmd, nullptr, false, true, true))
         {
             StringBuffer output;
+            StringBuffer outputErr;
             Owned<ISimpleReadStream> pipeReader = pipe->getOutputStream();
+            Owned<ISimpleReadStream> pipeReaderErr = pipe->getErrorStream();
             readSimpleStream(output, *pipeReader);
+            readSimpleStream(outputErr, *pipeReaderErr);
```
Comment (jakesmith) [inline system/jlib/jdebug.cpp]: I think :
```
StringBuffer outputErr;
Owned<ISimpleReadStream> pipeReaderErr = pipe->getErrorStream();
readSimpleStream(outputErr, *pipeReaderErr);
```

could be moved into the else.
i.e. delay until/if needed.

**Comment by jakesmith** [comment]
@mckellyln - looks good, 1 comment.

**Comment by mckellyln** [comment]
@jakesmith I thought those calls should be done before the wait() ?  I will double-check.

**Comment by mckellyln** [comment]
@jakesmith added 8k stderr buffer size and moved code into else block.
Because its such a small change I did an amend to this one commit.
Please re-review.

File: system/jlib/jdebug.cpp [hunk 2]
```diff
@@ -1995,7 +1995,7 @@ class CExtendedStats  // Disk network and cpu stats
         // and match those with entries in /proc/diskstats
         StringBuffer cmd("lsblk -o TYPE,MAJ:MIN --pairs");
         Owned<IPipeProcess> pipe = createPipeProcess();
-        if (pipe->run("list disks", cmd, nullptr, false, true))
+        if (pipe->run("list disks", cmd, nullptr, false, true, true, 8192))
```
Comment (jakesmith) [inline system/jlib/jdebug.cpp]: trivial: I'd consider putting constants like this into a #define or static const member, so it has a meaningful name.

**Comment by jakesmith** [comment]
@mckellyln - 1 v. minor comment - good to merge as is.

**Comment by mckellyln** [comment]
@richardkchapman ok to merge

**Comment by richardkchapman** [comment]
@mckellyln Not sure I understand this change. Comment suggests you are suppressing output of stderr information unless the command fails, code looks as if it is outputting stderr if command fails but would not have been previously?

**Comment by jakesmith** [comment]
Is it that as it was before (with haserror=false), that the pipe programs stderr is automatically part of the log because jlog usage redirects stderr to the log?

**Comment by mckellyln** [comment]
For components stderr was not logged.
For commands such as eclcc stderr went to console.
This PR captures stderr if command fails and then outputs it more appropriately.

**Pull request merged**